### DISCUSSION TITLE
Shell completions like fig but in the terminal completely

### DISCUSSION THREAD

**Discussion by dlvhdr** [charmbracelet/inspo#19] [Ideas]
Shell autocompletions UI is meh.

E.g., this is fish's:
![image](https://user-images.githubusercontent.com/6196971/199097928-871c16b1-7168-4e48-9cab-d51bc23f7790.png)
![image](https://user-images.githubusercontent.com/6196971/199098438-61479092-a791-4dab-b053-e119ee968487.png)

**My issues with it are:**
* They take the whole width of the terminal
* The description is very disconnected from the actual completion
* They show auto completions both horizontally and vertically.
* They don't have nerd font icons :(

**Describe the solution you'd like**
Something more sleek and modern looking, with nerd fonts maybe.
Fig nailed it, but they aren't CLI native and they hurt the startup time for new shell sessions.

![image](https://user-images.githubusercontent.com/6196971/199099081-18872071-8368-488f-9096-25e6a1960625.png)

Another example is nvim-cmp for Neovim.

We can (maybe?) even reuse the autocompletions by fig which are open source:
https://github.com/withfig/autocomplete

**Comment by aschey** [comment]
My project [bubbleprompt](https://github.com/aschey/bubbleprompt) might be relevant here. It started out as a port of [go-prompt](https://github.com/c-bata/go-prompt), but has since evolved into supporting more advanced features. Sounds like what you're describing is a way to have completions running directly in the terminal like a plugin, whereas bubbleprompt currently only runs as a typical TUI app. I've thought about this use case as well, but haven't really spent time exploring the options yet. Guessing it would either have to be something like a zsh plugin or maybe a custom prompt like powerlevel10k.

Bubbleprompt isn't quite ready for prime time yet as the public API needs some cleaning up and there are a couple more features from go-prompt that I want to port over, but here are a few quick examples:

[Interactive CLI](https://github.com/aschey/bubbleprompt-cobra/tree/main/examples/keyvalue) mostly autogenerated from a cobra app using [bubbleprompt-cobra](https://github.com/aschey/bubbleprompt-cobra):
![keyvalue](https://user-images.githubusercontent.com/5882266/200182855-624d445a-48a0-45eb-b44e-7b3289e3b265.gif)

[Simple javascript REPL](https://github.com/aschey/bubbleprompt/tree/main/examples/javascript) using a custom parser:
![javascript](https://user-images.githubusercontent.com/5882266/200183006-4e72319c-426d-41f8-9f3a-618f870f5744.gif)

**Comment by rsteube** [comment]
You might want to give [elvish](https://elv.sh/) a try.
